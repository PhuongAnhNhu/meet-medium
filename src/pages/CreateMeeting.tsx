import React from 'react';
import { Autocomplete, Box, Button, Chip, FormControl, FormGroup, TextField, Typography } from '@mui/material';
import AdapterDateFns from '@mui/lab/AdapterDateFns';
import LocalizationProvider from '@mui/lab/LocalizationProvider';
import { DatePicker } from '@mui/lab';
import { useSelector } from 'react-redux';
import { RootState } from 'store';
import { Room } from 'api/room';

const CreateMeeting = () => {
  const [value, setValue] = React.useState<Date | null>(new Date());
  const handleClick = () => {
    console.info('You clicked the Chip.');
  };
  const timeOptions = [{ label: '10 Min' }, { label: '20 Min' }, { label: '30 Min' }];
  const roomInBerlin = [
    { label: 'Berlin - 7.01 Catan', address: 'Berlin-701Catan@endava.com' },
    { label: 'Berlin - 7.02 London', address: 'Berlin-702London@endava.com' },
    { label: 'Berlin - 7.03 Aquarium', address: 'Berlin-703Aquarium@endava.com' },
    { label: 'Berlin - 7.04 Paris', address: 'Berlin-704Paris@endava.com' },
    { label: 'Berlin - 7.05 Telefonzelle', address: 'Berlin-705Tefeonzelle@endava.com' },
    { label: 'Berlin - 7.06 Bonn', address: 'Berlin-706Bonn@endava.com' },
    { label: 'Berlin - 7.07 Moskau', address: 'Berlin-707Moskau@endava.com' },
  ];
  return (
    <Box mt={4} mr={2}>
      <FormGroup sx={{ width: '100%' }}>
        <Typography variant="h3">Raum Buchen</Typography>
        <FormControl margin="dense" fullWidth>
          <LocalizationProvider dateAdapter={AdapterDateFns}>
            <DatePicker
              label="Basic example"
              value={value}
              onChange={(newValue) => {
                setValue(newValue);
              }}
              renderInput={(params) => <TextField {...params} />}
            />
          </LocalizationProvider>
        </FormControl>
        <FormControl margin="dense" fullWidth>
          <Autocomplete
            disablePortal
            id="combo-box-demo"
            options={timeOptions}
            renderInput={(params) => <TextField {...params} label="Dauert" />}
          />
        </FormControl>
        <FormControl margin="dense" fullWidth>
          <Autocomplete
            disablePortal
            id="combo-box-demo"
            options={roomInBerlin}
            renderInput={(params) => <TextField {...params} label="Raum" />}
          />
        </FormControl>
        <Box mb={2} mt={2} sx={{ display: 'flex', flexFlow: 'row' }}>
          <Chip label="10:00 - 20:00" onClick={handleClick} />
          <Chip label="15:20 - 16:40" onClick={handleClick} />
        </Box>

        <Button variant="contained">Buchen</Button>
      </FormGroup>
    </Box>
  );
};

export default CreateMeeting;
