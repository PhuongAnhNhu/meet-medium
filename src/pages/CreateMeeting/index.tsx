import React, { useEffect, useState, FormEvent, SyntheticEvent } from 'react';
import { RootState, useAppDispatch } from 'store';
import {
  Autocomplete,
  Box,
  Button,
  FormControl,
  FormControlLabel,
  Radio,
  RadioGroup,
  TextField,
  Typography,
  CircularProgress,
  Snackbar,
  Alert,
} from '@mui/material';
import AdapterDateFns from '@mui/lab/AdapterDateFns';
import LocalizationProvider from '@mui/lab/LocalizationProvider';
import { DateTimePicker } from '@mui/lab';
import { useSelector } from 'react-redux';
import { getRoomOptions, getTimeOptions } from '../../helper/suggestion';
import { createEvent, findMeetingsTime } from 'store/features/roomSlice';
import { format } from 'date-fns';

const TIME_FORMAT = 'HH:mm';

const CreateMeeting = () => {
  const initialFormState: MeetingForm = {
    datetime: new Date(),
    period: '',
    room: '',
    timeslot: [],
  };

  const [formState, setFormState] = useState<MeetingForm>(initialFormState);
  const [roomOptions, setRoomOptions] = useState<string[]>([]);
  const [timeOptions, setTimeOptions] = useState<string[][]>([]);
  const [open, setOpen] = useState<boolean>(false);

  const { creating, created } = useSelector((state: RootState) => state.room);

  /**
   * Only used to get the timeOptions and the roomOptions
   */
  const allMeetingData = useSelector((state: RootState) => state.room.meetingTimeSuggestion);

  const dispatch = useAppDispatch();

  useEffect(() => {
    if (allMeetingData.length > 0) {
      const roomOptions = getRoomOptions(allMeetingData);
      setRoomOptions(roomOptions);
    }
  }, [allMeetingData]);

  useEffect(() => {
    setOpen(created);
    if (created) {
      setFormState(initialFormState);
      setTimeOptions([]);
      setRoomOptions([]);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [created]);

  const handleDateTimeChange = (date: Date | null) => {
    date && setFormState((currentFormState) => ({ ...initialFormState, datetime: date }));
  };

  const handleChangePeriod = (event: React.ChangeEvent<HTMLInputElement>) => {
    const period = event.target.value;
    setFormState((currentFormState) => ({ ...currentFormState, period, room: '' }));
    if (period) {
      dispatch(findMeetingsTime({ datetime: formState.datetime, period }));
    }
    if (!period) {
      setTimeOptions([]);
    }
  };

  const handleRoomChange = (event: SyntheticEvent<Element, Event>, room: string | null) => {
    if (room) {
      setFormState((currentFormState) => ({ ...currentFormState, room }));
      const timeOptions = getTimeOptions(room, allMeetingData);
      setTimeOptions(timeOptions);
    }
  };

  const handleTimeslotChange = (event: any) => {
    setFormState((currentFormState) => ({
      ...currentFormState,
      timeslot: timeOptions[event.target.value],
    }));
  };

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    dispatch(createEvent(formState));
    debugger;
  };
  return (
    <Box mt={4} sx={{ width: '99%' }}>
      <form onSubmit={handleSubmit}>
        <Typography variant="h3">Raum Buchen</Typography>
        <FormControl margin="dense" fullWidth>
          <LocalizationProvider dateAdapter={AdapterDateFns}>
            <DateTimePicker
              ampm={false}
              label="Datum"
              value={formState.datetime}
              onChange={handleDateTimeChange}
              renderInput={(params) => <TextField {...params} name="datetime" />}
            />
          </LocalizationProvider>
        </FormControl>

        <FormControl margin="dense" fullWidth>
          <TextField
            id="outlined-basic"
            label="Dauer"
            type="number"
            variant="outlined"
            value={formState.period}
            onChange={handleChangePeriod}
            name="period"
          />
        </FormControl>

        <FormControl margin="dense" fullWidth>
          <Autocomplete
            value={formState.room}
            disablePortal
            options={[...roomOptions, '']}
            disabled={!(formState.period && formState.datetime)}
            onChange={handleRoomChange}
            renderInput={(params) => <TextField {...params} label="Raum" name="roomname" />}
          />
        </FormControl>

        {!!timeOptions && (
          <Box mb={2} mt={2} sx={{ display: 'flex', flexFlow: 'row' }}>
            <FormControl>
              <RadioGroup name="timeslot" onChange={handleTimeslotChange}>
                {timeOptions?.map((item, key) => {
                  return (
                    <FormControlLabel
                      key={key}
                      value={key}
                      control={<Radio />}
                      label={`${format(new Date(item[0]), TIME_FORMAT)}-${format(new Date(item[1]), TIME_FORMAT)}`}
                    />
                  );
                })}
              </RadioGroup>
            </FormControl>
          </Box>
        )}

        <Button type="submit" variant="contained">
          {creating ? <CircularProgress size="1rem" color="secondary" /> : 'Buchen'}
        </Button>
      </form>

      <Snackbar open={open} autoHideDuration={3000} onClose={() => setOpen(false)}>
        <Alert severity="success" sx={{ width: '100%' }}>
          Der Raum wurde gebucht!
        </Alert>
      </Snackbar>
    </Box>
  );
};

export default CreateMeeting;
