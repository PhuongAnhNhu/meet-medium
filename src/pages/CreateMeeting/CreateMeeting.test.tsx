import { render } from '@testing-library/react';
import { wrapWithStore } from 'tests/wrappers';
import CreateMeeting from '.';

describe('CreateMeeting', () => {
  beforeEach(() => {
    jest.useFakeTimers().setSystemTime(new Date('2022-05-22'));
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  it('renders', () => {
    const { container } = render(wrapWithStore(<CreateMeeting />));

    expect(container).toMatchSnapshot();
  });
});
