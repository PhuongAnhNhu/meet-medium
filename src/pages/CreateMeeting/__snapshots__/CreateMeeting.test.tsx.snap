// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMeeting renders 1`] = `
<div>
  <div
    class="MuiBox-root css-biotn8"
  >
    <form>
      <h3
        class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
      >
        Raum Buchen
      </h3>
      <div
        class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="mui-1"
            id="mui-1-label"
          >
            Datum
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              aria-label="Choose date, selected date is May 22, 2022"
              aria-readonly="true"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="mui-1"
              name="datetime"
              readonly=""
              type="text"
              value="05/22/2022 16:28"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1z7n62"
              >
                <span>
                  Datum
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="outlined-basic"
            id="outlined-basic-label"
          >
            Dauer
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id="outlined-basic"
              name="period"
              type="number"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-173wfxe"
              >
                <span>
                  Dauer
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
      >
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
          role="combobox"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-disabled css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="mui-4"
              id="mui-4-label"
            >
              Raum
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiOutlinedInput-input MuiInputBase-input Mui-disabled MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                disabled=""
                id="mui-4"
                name="roomname"
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                  disabled=""
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ArrowDropDownIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M7 10l5 5 5-5z"
                    />
                  </svg>
                </button>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-173wfxe"
                >
                  <span>
                    Raum
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiBox-root css-oazcq0"
      >
        <div
          class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
        >
          <div
            class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
            role="radiogroup"
          />
        </div>
      </div>
      <button
        class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Buchen
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</div>
`;
